proof/verifier.go

```diff
@@ -184,67 +257,232 @@ func (p *Proof) verifySplitRootProof() error {
-		outputCommitVersion := commitVersion
-		if !commitment.IsSimilarTapCommitmentVersion(
-			&firstCommitVersion, &outputCommitVersion,
-		) {
+	// Create a P2TROutputsSTXOs from p2trOutputs and map STXOs by
+	// serialized key.
+	assetMap := make(map[asset.SerializedKey]*asset.Asset)
+	p2trOutputsSTXOs := make(P2TROutputsSTXOs)
+	for outIdx := range p2trOutputs {
+		p2trOutputsSTXOs[outIdx] = make(fn.Set[asset.SerializedKey])
+	}
+
+	for idx := range stxoAssets {
+		stxoAsset := stxoAssets[idx].(*asset.Asset)
+		key := asset.ToSerialized(stxoAsset.ScriptKey.PubKey)
+		assetMap[key] = stxoAsset
+		for outIdx := range p2trOutputs {
+			p2trOutputsSTXOs[outIdx].Add(key)
+		}
+	}
+
+	for idx := range p.ExclusionProofs {
+		exclusionProof := p.ExclusionProofs[idx]
+		if exclusionProof.TapscriptProof != nil &&
+			exclusionProof.TapscriptProof.Bip86 {
+
+			continue
+		}
 
-			log.Tracef("output %d commit version %d, first output "+
-				"commit version %d", outputIdx, commitVersion,
-				firstCommitVersion)
+		err := verifySTXOProofSet(
+			&p.AnchorTx, exclusionProof, assetMap, p2trOutputsSTXOs,
+			false,
+		)
+		if err != nil {
+			return err
+		}
+	}
 
-			return nil, fmt.Errorf("mixed anchor commitment " +
-				"versions for exclusion proofs")
+	return nil
+}
+
+// verifySTXOProofSet verifies a set of STXO proofs.
+func verifySTXOProofSet(anchorTx *wire.MsgTx, baseProof TaprootProof,
+	assetMap map[asset.SerializedKey]*asset.Asset,
+	p2trOutputs P2TROutputsSTXOs,
+	inclusion bool) error {
+
+	for key := range baseProof.CommitmentProof.STXOProofs {
+		stxoProof := baseProof.CommitmentProof.STXOProofs[key]
+		stxoAsset := assetMap[key]
+		stxoCombinedProof := MakeSTXOProof(baseProof, &stxoProof)
+
+		_, err := verifyTaprootProof(
+			anchorTx, &stxoCombinedProof, stxoAsset, inclusion,
+		)
+		if err != nil {
+			return fmt.Errorf("error verifying STXO proof: %w",
+				err)
+		}
+
+		outIdx := stxoCombinedProof.OutputIndex
+		delete(p2trOutputs[outIdx], key)
+		if len(p2trOutputs[outIdx]) == 0 {
+			delete(p2trOutputs, outIdx)
+		}
+	}
+
+	return nil
+}
+
+// MakeSTXOProof creates a new proof for an STXO reusing the base proof while
+// replacing commitmentProof but the stxoProof.
+func MakeSTXOProof(baseProof TaprootProof,
+	stxoProof *commitment.Proof) TaprootProof {
+
+	return TaprootProof{
+		OutputIndex: baseProof.OutputIndex,
+		InternalKey: baseProof.InternalKey,
+		CommitmentProof: &CommitmentProof{
+			Proof: commitment.Proof{
+				TaprootAssetProof: stxoProof.TaprootAssetProof,
+				AssetProof:        stxoProof.AssetProof,
+				UnknownOddTypes:   stxoProof.UnknownOddTypes,
+			},
+			TapSiblingPreimage: baseProof.CommitmentProof.
+				TapSiblingPreimage,
+		},
+		TapscriptProof:  baseProof.TapscriptProof,
+		UnknownOddTypes: baseProof.UnknownOddTypes,
+	}
+}
+
+// verifySTXOVersions verifies all STXO versions match.
+func verifySTXOVersions(versions map[uint32][]commitment.TapCommitmentVersion) (
```

Comment:
IIUC, this isn't STXO specific, it's generic to the asset commitment: we want to make sure all the asset commitments have the same version. Looks like this was refactored out of the old logic, so I think we can just rename it to: `assertUniformCommitVersions` or something like that?